go/def.bzl

```diff
@@ -999,9 +999,172 @@ def cgo_library(name, srcs,
+    rev = ctx.attr.commit
+  elif ctx.attr.tag:
+    rev = ctx.attr.tag
+  else:
+    fail("neither commit or tag is specified", "commit")
+
+  result = ctx.execute([
+      fetch_repo,
+      '--dest', ctx.path(''),
+      '--remote', ctx.attr.importpath,
+      '--rev', rev])
+  if result.return_code:
+    fail("failed to fetch %s: %s" % (ctx.attr.importpath, result.stderr))
+
+def _new_go_repository_impl(ctx):
+  _go_repository_impl(ctx)
+  gazelle = ctx.path(ctx.attr._gazelle)
+
+  result = ctx.execute([
+      gazelle,
+      '--go_prefix', ctx.attr.importpath, '--mode', 'fix',
+      ctx.path('')])
+  if result.return_code:
+    fail("failed to generate BUILD files for %s: %s" % (
+        ctx.attr.importpath, result.stderr))
+
+_go_repository_attrs = {
+    "importpath": attr.string(mandatory = True),
+    "commit": attr.string(),
+    "tag": attr.string(),
+
+    "_fetch_repo": attr.label(
+        default = Label("@io_bazel_rules_go_repository_tools//:bin/fetch_repo"),
+        allow_files = True,
+        single_file = True,
+    ),
+}
+
+go_repository = repository_rule(
+    _go_repository_impl,
+    attrs = _go_repository_attrs,
+)
+
+new_go_repository = repository_rule(
+    _new_go_repository_impl,
+    attrs = _go_repository_attrs + {
+        "_gazelle": attr.label(
+            default = Label("@io_bazel_rules_go_repository_tools//:bin/gazelle"),
+            allow_files = True,
+            single_file = True,
+        ),
+    },
+)
 
 ################
 
+repository_tool_deps = {
+    'buildifier': struct(
+        importpath = 'github.com/bazelbuild/buildifier',
+        repo = 'https://github.com/bazelbuild/buildifier',
+        commit = '0ca1d7991357ae7a7555589af88930d82cf07c0a',
+    ),
+    'tools': struct(
+        importpath = 'golang.org/x/tools',
+        repo = 'https://github.com/golang/tools',
+        commit = '2bbdb4568e161d12394da43e88b384c6be63928b',
+    )
+}
+
+def go_internal_tools_deps():
+  """only for internal use in rules_go"""
+  go_repository(
+      name = "io_bazel_buildifier",
+      commit = repository_tool_deps['buildifier'].commit,
+      importpath = repository_tool_deps['buildifier'].importpath,
+  )
+
+  new_go_repository(
+      name = "org_golang_x_tools",
+      commit = repository_tool_deps['tools'].commit,
+      importpath = repository_tool_deps['tools'].importpath,
+  )
+
+def _fetch_repository_tools_deps(ctx, goroot, gopath):
+  for name, dep in repository_tool_deps.items():
+    result = ctx.execute(['mkdir', '-p', ctx.path('src/' + dep.importpath)])
+    if result.return_code:
+      fail('failed to create directory: %s' % result.stderr)
+    ctx.download_and_extract(
+        '%s/archive/%s.zip' % (dep.repo, dep.commit),
+        'src/%s' % dep.importpath, '', 'zip', '%s-%s' % (name, dep.commit))
+
+  result = ctx.execute([
+      'env', 'GOROOT=%s' % goroot, 'GOPATH=%s' % gopath, 'PATH=%s/bin' % goroot,
+      'go', 'generate', 'github.com/bazelbuild/buildifier/core'])
```

Comment:
No. Here we need to build `gazelle` and `fetch_repo` without any help by Bazel build rules because the tools are built and used in the loading phase.
This is a variation of the typical issues of bootstrapping.